File: RoundRobin.cpp
#include <iostream>
#include <vector>
#include "FCFS.h"

#define QUANTUM 2

class RoundRobin: public FCFS {
    protected:
        int* remBurstTime;
        int currentTime = 0;

    public:
        void calculate() {//Overriding void FCFS::calculate()
            for (int i = 0 ; i < numOfProcesses ; i++)
                remBurstTime[i] = burstTime[i];
        
            int t = 0; // Current time
        
            // Keep traversing processes in round robin manner
            // until all of them are not done.
            while (1)
            {
                bool done = true;
        
                // Traverse all processes one by one repeatedly
                for (int i = 0 ; i < numOfProcesses; i++)
                {
                    // If burst time of a process is greater than 0
                    // then only need to process further
                    if (remBurstTime[i] > 0)
                    {
                        done = false; // There is a pending process
        
                        if (remBurstTime[i] > QUANTUM)
                        {
                            // Increase the value of t i.e. shows
                            // how much time a process has been processed
                            t += QUANTUM;
        
                            // Decrease the burst_time of current process
                            // by QUANTUM
                            remBurstTime[i] -= QUANTUM;
                        }
        
                        // If burst time is smaller than or equal to
                        // QUANTUM. Last cycle for this process
                        else
                        {
                            // Increase the value of t i.e. shows
                            // how much time a process has been processed
                            t = t + remBurstTime[i];
        
                            // Waiting time is current time minus time
                            // used by this process
                            waitingTime[i] = t - burstTime[i];
        
                            // As the process gets fully executed
                            // make its remaining burst time = 0
                            remBurstTime[i] = 0;
                        }
                    }
                }
        
                // If all processes are done
                if (done == true)
                break;
            }
        }

        RoundRobin(): FCFS() {
            remBurstTime = new int[numOfProcesses];
        }

        ~RoundRobin() {
            delete[] remBurstTime;
        }
};

int main() {
    RoundRobin obj1;
    obj1.fetchData();
    obj1.calculate();
    obj1.printResults();

    return 0;
}

File: RoundRobin.java
import java.util.Scanner;

public class RoundRobin {
    int numOfProcesses;
    int timeQuantum;
    int sumOfQuantum = 0;
    int temp;
    int count = 0;
    int[] burstTime, waitingTime, turnAroundTime, remainingTime;
    private static final Scanner sc = new Scanner(System.in);

    void fetchData() {
        for (int i = 0; i < numOfProcesses; i++) {
            System.out.print("Enter the burst time for Process " + (i + 1) + ": ");
            remainingTime[i] = burstTime[i] = sc.nextInt();
        }
    }

    void calculate() {
        while(count != numOfProcesses) {
            for(int i = 0; i < numOfProcesses; i++) {
                if(remainingTime[i] == 0) {
                    count++;
                    continue;
                }
                temp = timeQuantum;
                if(remainingTime[i] > timeQuantum) {
                    remainingTime[i] -= timeQuantum;
                } else if(remainingTime[i] >= 0) {
                    temp = remainingTime[i];
                    remainingTime[i] = 0;
                    turnAroundTime[i] = sumOfQuantum + temp;
                }
                sumOfQuantum += temp;
                // System.out.println("Process | Burst Time | Remaining BT");
                // System.out.println((i + 1) + "\t\t" + burstTime[i]+ "\t\t" + remainingTime[i]);
                // System.out.println("-----------------------------------------------------");
                // System.out.println("Current Time: " + sumOfQuantum + "\n" + "Loop Iteration " + i + "\n" + "Count: " + count);
            }
        }
        for (int i = 0; i < numOfProcesses; i++) {
            waitingTime[i] = turnAroundTime[i] - burstTime[i];
        }
    }

    void printResults() {
        System.out.println();
        System.out.println("Process | Burst Time | Turnaround Time | Waiting Time");
        for(int i = 0; i < numOfProcesses; i++) {
            System.out.println((i + 1) + "\t\t" + burstTime[i]+ "\t\t" + turnAroundTime[i] + "\t\t" + waitingTime[i]);
            System.out.println("-----------------------------------------------------");
        }
    }

    RoundRobin() {
        System.out.print("Enter the number of processes: ");
        numOfProcesses = sc.nextInt();
        System.out.print("Enter Time Quantum: ");
        timeQuantum = sc.nextInt();
        System.out.println();

        burstTime = new int[numOfProcesses];
        waitingTime = new int[numOfProcesses];
        turnAroundTime = new int[numOfProcesses];
        remainingTime = new int[numOfProcesses];
    }

    public static void main(String[] args) {
        RoundRobin obj1 = new RoundRobin();

        obj1.fetchData();
        obj1.calculate();
        obj1.printResults();
    }
}
